### PR TITLE
Swappable models

### LINKED ISSUE #306: review migrations in virtualenv
btw since I want to add fields on one of the model, I had a pb with a migration: when executing ./manage.py makemigrations it generated a migration for the schedule app, i.e. directly in my virtual env (that kind of things doesn't play well when rolling out to CI or production...). I was able to tweak it with the following, but it's not very optimal. Do you have plan to make things better? I can definitely write some code as well:

class CrossAppAddField(migrations.AddField):
    """
    Only way I found to migrate the model of another app.
    """
    def __init__(app_label, *args, **kwargs):
        self.app_label = app_label
        super(CrossAppAddField, self).__init__(*args, **kwargs)

    def state_forwards(self, app_label, *args, **kwargs):
        """
        Override app_label.
        """
        return super(CrossAppAddField, self).state_forwards(self.app_label, *args, **kwargs)

    def database_forwards(self, app_label, *args, **kwargs):
        """
        Override app_label.
        """
        return super(CrossAppAddField, self).database_forwards(self.app_label, *args, **kwargs)

    def database_backwards(self, app_label, *args, **kwargs):  # pragma: no cover
        """
        Override app_label.
        """
        return super(CrossAppAddField, self).database_backwards(self.app_label, *args, **kwargs)
And then in one of my migrations, I would do: CrossAppAddField('schedule', model_name='event', name='event_type', field=models.CharField(...))

### PR DESCRIPTION
**Pull request by m-vdb**
Fixes #306

This PR aims at proposing a powerful replacement for base classes that were removed in #377.
⚠️ Do not merge for now, we need to test the version change on live projects

Changes summary:

- split models between abstract models and actual models
- define settings for swappable models + getters (*à la `get_user_model()`*)
- add Django migrations (❗️this point needs to be validated and tested with live projects)
- add tests for the whole thing

### CODE REVIEW

**Comment by jdufresne** [comment]
Do you mind cleaning up the lint errors so the Travis tests will run?

Do we really need a to allow swapping _all_ models? Or would handling just `Event` be sufficient? I'd prefer not to add additional complexity if it isn't handling a real use case.

**Comment by m-vdb** [comment]
Lint should be good after 581d0c9 and 3debd40. About swapping all the models: in my use cases, I have custom models for `Event` and `Occurrence`, but not for `Calendar`. I would argue that it is pretty useful to have this functionality for the occurrences as well. The swappability for the `Calendar` is not a must have in my opinion, but in some cases developers could have several calendar per user (for instance) and would like to enrich the calendar model. Since it was the same amount of work I decided to do it, but I can easily revert that.

I still need to figure out how to fix tests + migrations, but what do you think about this PR?

**Comment by llazzaro** [comment]
@m-vdb don't upgrade to 0.8.7 or do it in test server. we just removed this feature in 0.8.7

**Comment by llazzaro** [comment]
@jdufresne Let review this change, we could be breaking other people code.
Let me know what you think about this PR.
I will try to make some time at night to review this PR

**Comment by jdufresne** [comment]
I'll try to find some time to give this a deeper review.

**Comment by keeth** [comment]
I'm very interested in this as I use base classes heavily, and I find it hard to maintain.

**Comment by jdufresne** [comment]
@m-vdb Can you rebase this PR so Travis runs against the latest mater? I'd like to see CI green before the review continues.

**Comment by m-vdb** [comment]
my branch starts from `develop`, isn't it the way to go? Like I said in a previous comment, I still need to figure out how to make my test pass and test that in a live project. I am not sure about the way I handled the migrations.

**Comment by jdufresne** [comment]
Oh you're right. Sorry, I misspoke. I intended to ask to rebase on develop. When you have a chance, can you do that and squash the commits? It would help keep the change tidy and aid reviewing.

Nice, I think that the general approach looks good! I think the use of swappable will be easier to maintain than the previous base classes.

I'd like to see more tests and have them passing before this is merge ready. The tests go a long way towards allowing me to analyze the change. Let me know when that is ready and I'll continue the review. It might help to look at how Django tests the swappable `User` model as guidance.

Nice work!

File: tests/test_swapped_models.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    """
+
+    def _get_migration_class(self):
+        try:
+            module = importlib.import_module('tests.test_project.migrations.0001_initial')
+            return module.Migration
+        except ImportError:
+            raise RuntimeError(self.__doc__.strip())
+
+    @override_settings(SCHEDULER_CALENDAR_MODEL='schedule.Calendar', SCHEDULER_EVENT_MODEL='schedule.Event')
```
Comment (jdufresne) [inline tests/test_swapped_models.py]: Why is this overridden to not be the alternate model? Shouldn't the new tests use the `CustomEvent` class?

File: .travis.yml [hunk 2]
```diff
@@ -27,6 +27,7 @@ before_script:
   - isort --check-only --diff
   - python -m django check --settings=tests.settings
   - python -m django makemigrations --settings=tests.settings --no-input --dry-run --check
+  - python -m django makemigrations test_project --settings=tests.test_project.settings
```
Comment (jdufresne) [inline .travis.yml]: This feels a bit odd to me on the surface. Could this instead be a part of running the test? I'd like to avoid an extra test setup step if possible so that the tests are immediately runnable as before.

Comment (m-vdb) [reply .travis.yml]: ok I'll see what I can do, I just wanted to avoid executing a script within the tests. but maybe I can just use `call_command()`, will try.

File: tests/test_swapped_models.py (hunk 1 above)
Comment (m-vdb) [reply tests/test_swapped_models.py]: I had an error with it but it might be related to my migration issue

Comment (m-vdb) [reply tests/test_swapped_models.py]: ah I know why:
- the tests are executed within the context of the `tests` project, using the `tests.settings` module
- the migrations is meaningful only in the context of the `tests.test_project`, using the `tests.test_project.settings`
- as a consequence, importing the migration file like I do in the tests results in an import error: the `django.conf.settings.SCHEDULER_CALENDAR_MODEL` isn't defined for `tests.settings` but it is defined for `tests.test_project.settings`

**Comment by jdufresne** [comment]
To resolve conflicts, do you mind rebasing instead of merging? It would make the commit history much cleaner and easier to review. Thanks.

**Comment by m-vdb** [comment]
done, sorry I missed that earlier. Here is where I am stuck: you can see that one of the tests that I introduced isn't passing. It is because I still haven't figured out how to migrate from the current version of `django-scheduler` to this one. Let me explain: if a user is under `django-scheduler==0.8.8` and uses `BASE_CLASSES` and wants to switch to this new version, he will need to swap the models. And here comes the migration issue (cf. backtraces on travis).

Also, I realized something interesting. Once you define the swapped models, there are two migrations that are generated: [one in the `schedule` module](https://github.com/llazzaro/django-scheduler/pull/389/files#diff-b6acd87a5a4d1d245edd00c0af2ee9b9), and one in the `test_project` module.

I've tried to search Django documentation and other projects GH issues and I've found similar issues, but I still don't know exactly how to fix that:
- https://github.com/villoid/django-oauth-toolkit/commit/4ef90936ff257d1b0d77c0631874ab9cfdd88d2c : warns about the migration when you swap a model. (when looking at [their models](https://github.com/evonove/django-oauth-toolkit/blob/master/oauth2_provider/models.py), they are in the same situation as here: several swappable models with foreignkeys, I'll dig deeper)
- https://github.com/wq/django-swappable-models/issues/12 : looks like exactly the same issue

**Comment by m-vdb** [comment]
I've managed to make it somehow work, by following [`django-oauth-toolkit` documentation](https://django-oauth-toolkit.readthedocs.io/en/latest/advanced_topics.html#extending-the-application-model). What users would have to do:
- define a custom model *without* extending the abstract class
- generate the migration
- define the swapped settings and make the custom model inherit from the abstract class
- generate a second migration: this one is a bit problematic because Django sees that you are adding new fields to the custom models, and some fields are defined non-nullable on the abstract classes, so it will prompt the user to provide one-off defaults. It's a small detail but in terms of UX it is a bit distressing:
```
You are trying to add a non-nullable field 'name' to customcalendar without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> 'stuff'
You are trying to add a non-nullable field 'slug' to customcalendar without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> 'stuff'
You are trying to add a non-nullable field 'calendar' to customevent without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
You are trying to add the field 'created_on' with 'auto_now_add=True' to customevent without a default; the database needs something to populate existing rows.

 1) Provide a one-off default now (will be set on all existing rows)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
You can accept the default 'timezone.now' by pressing 'Enter' or you can provide another value.
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
[default: timezone.now] >>> timezone.now()
You are trying to add a non-nullable field 'end' to customevent without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
You are trying to add a non-nullable field 'start' to customevent without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
You are trying to add a non-nullable field 'title' to customevent without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
You are trying to add the field 'created_on' with 'auto_now_add=True' to customoccurrence without a default; the database needs something to populate existing rows.

 1) Provide a one-off default now (will be set on all existing rows)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
You can accept the default 'timezone.now' by pressing 'Enter' or you can provide another value.
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
[default: timezone.now] >>> None
You are trying to add a non-nullable field 'end' to customoccurrence without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
You are trying to add a non-nullable field 'event' to customoccurrence without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
You are trying to add a non-nullable field 'original_end' to customoccurrence without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
You are trying to add a non-nullable field 'original_start' to customoccurrence without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
You are trying to add a non-nullable field 'start' to customoccurrence without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> None
Migrations for 'test_project':
  tests/test_project/migrations/0002_auto_20180116_0715.py
    - Change Meta options on customcalendar
    - Change Meta options on customevent
    - Change Meta options on customoccurrence
    - Add field name to customcalendar
    - Add field slug to customcalendar
    - Add field calendar to customevent
    - Add field color_event to customevent
    - Add field created_on to customevent
    - Add field creator to customevent
    - Add field description to customevent
    - Add field end to customevent
    - Add field end_recurring_period to customevent
    - Add field rule to customevent
    - Add field start to customevent
    - Add field title to customevent
    - Add field updated_on to customevent
    - Add field cancelled to customoccurrence
    - Add field created_on to customoccurrence
    - Add field description to customoccurrence
    - Add field end to customoccurrence
    - Add field event to customoccurrence
    - Add field original_end to customoccurrence
    - Add field original_start to customoccurrence
    - Add field start to customoccurrence
    - Add field title to customoccurrence
    - Add field updated_on to customoccurrence
    - Alter index_together for customevent (1 constraint(s))
    - Alter index_together for customoccurrence (1 constraint(s))
```

**Comment by m-vdb** [comment]
@jdufresne ok I think I'm good for all logical aspects, even for the migration. Tests are green now, can you take a look? Also, I think that this requires documentation and an extended migration guide to this new version; it shouldn't be too hard to write, I can do it before we merge this.

File: tests/test_swapped_models.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    This test verifies that it is possible to install `django-scheduler`
+    and then swap its models.
+    """
+    settings_path = os.path.join(settings.BASE_DIR, 'tests', 'test_project', 'settings.py')
+    updated_settings_path = os.path.join(settings.BASE_DIR, 'tests', 'test_project', 'updated_settings.py.tpl')
+
+    def setUp(self):
+        super(TestSwappableModels, self).setUp()
+        # cleanup any possible leftovers before running
+        self._cleanup()
```
Comment (jdufresne) [inline tests/test_swapped_models.py]: Why do you need to call `_cleanup` both before and after? Shouldn't one be sufficient?

File: tests/settings.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Build paths inside the project like this: os.path.join(BASE_DIR, ...)
 import os
 
-BASE_DIR = os.path.dirname(os.path.dirname(__file__))
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
```
Comment (jdufresne) [inline tests/settings.py]: Was this change required to make the tests work? Can you explain?

File: .gitignore [hunk 5]
```diff
@@ -9,3 +9,5 @@
 /dist/
 *.egg-info/
 /docs/_build/
+# generated in tests, shouldn't be versionned
+tests/test_project/migrations/0001_initial.py
```
Comment (jdufresne) [inline .gitignore]: I don't understand why this is in gitignore. Shouldn't this exist as a normal file?

File: tests/test_project/settings.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
+# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
+import os
+
+BASE_DIR = os.path.dirname(os.path.dirname(__file__))
+
+SECRET_KEY = '_nc9_+ntllx!ptq5tzh96nn^$a#81#auhrdw#8t#+n0#4ig(lz'
```
Comment (jdufresne) [inline tests/test_project/settings.py]: Per 8343ddeeb157f514cad195fd8a0e4ea25d249cdd, please use an obviously fake secret for tests.

File: tests/test_swapped_models.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                )
+            )
+
+    def _execute_makemigrations(self):
+        """
+        Execute `python -m django makemigrations test_project --settings=tests.test_project.settings`.
+        """
+        popen = subprocess.Popen(
+            [
+                'python',
```
Comment (jdufresne) [inline tests/test_swapped_models.py]: This should probably use [`sys.executable`](https://docs.python.org/3/library/sys.html#sys.executable)

File: tests/test_swapped_models.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+                '-m',
+                'django',
+                'makemigrations',
+                'test_project',
+                '--settings=tests.test_project.settings'
+            ],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
```
Comment (jdufresne) [inline tests/test_swapped_models.py]: As you're not using the input/output, you can use [`DEVNULL`](https://docs.python.org/3/library/subprocess.html#subprocess.DEVNULL) instead of `PIPE`.

File: tests/settings.py (hunk 4 above)
Comment (m-vdb) [reply tests/settings.py]: see 8c45306 : `__file__` is only absolute in python3, so it was never defined right in python2. This had no consequences before because this settings wasn't used throughout the tests but now it is.

File: tests/test_project/settings.py (hunk 6 above)
Comment (m-vdb) [reply tests/test_project/settings.py]: 8fd293d ✅

File: tests/test_swapped_models.py (hunk 3 above)
Comment (m-vdb) [reply tests/test_swapped_models.py]: if for some reason the test completely crashes, this makes sure the test "setup" isn't corrupted. I can easily remove, it is just a nice-to-have for the developers of this repo.

File: tests/test_swapped_models.py (hunk 7 above)
Comment (m-vdb) [reply tests/test_swapped_models.py]: 9661a38 ✅

File: tests/test_swapped_models.py (hunk 8 above)
Comment (m-vdb) [reply tests/test_swapped_models.py]: in `self._check_command_return()` I do. In case the command exited with a non-zero return code, this method will display the `stdout` and `stderr` of the command, see:

https://github.com/m-vdb/django-scheduler/blob/8c45306aec5d2b8b1592f7c327cecfc6bb2558a7/tests/test_swapped_models.py#L52-L69

File: .gitignore (hunk 5 above)
Comment (m-vdb) [reply .gitignore]: the whole purpose of the test I wrote is to generate this file on the fly and verify that it is well formatted. If for some reason in the future someone inadvertently alters the shape of the generated file by changing the code of swappable models, then we will know it and the unit tests will break. Versioning this file doesn't make sense because it is generated by the tests. To avoid confusion for this repo's collaborators I chose to add this file to gitignore. If the test ever fails and the cleanup phase of the test fails to execute too, this file won't be visible with `git status` and a collaborator won't risk versioning it.

**Comment by thongly** [comment]
@jdufresne Is this still awaiting anything before a merge?

The base class is key, in my opinion, to address the problem of extensibility while avoiding the problems #377

**Comment by thongly** [comment]
@m-vdb - I am testing a few things with this branch and tied inheriting from the various `AbstractBaseClass`es.  

I noticed that I am getting a reverse accessor name clash.  However, if the special related_name syntax is used, this goes away - [as per this section in the Django docs](https://docs.djangoproject.com/en/2.0/topics/db/models/#be-careful-with-related-name-and-related-query-name)

`calendar = models.ForeignKey(
        SCHEDULER_CALENDAR_MODEL,
        on_delete=models.CASCADE,
        verbose_name=_("calendar"),
    related_name='%(class)s'        
        )`

**Comment by thongly** [comment]
Onto one of the final steps now, and I noticed that one of the migrations seems not to recognize the derived name of the `Calendar` class from `settings.py`.

> Running migrations:
  Applying schedule.0001_initial...Traceback (most recent call last):
  File "manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 365, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 335, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 200, in handle
    fake_initial=fake_initial,
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/migration.py", line 122, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/operations/models.py", line 92, in database_forwards
    schema_editor.create_model(model)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/base/schema.py", line 249, in create_model
    definition, extra_params = self.column_sql(model, field)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/base/schema.py", line 141, in column_sql
    db_params = field.db_parameters(connection=self.connection)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/related.py", line 966, in db_parameters
    return {"type": self.db_type(connection), "check": self.db_check(connection)}
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/related.py", line 963, in db_type
    return self.target_field.rel_db_type(connection=connection)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/related.py", line 877, in target_field
    return self.foreign_related_fields[0]
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/related.py", line 634, in foreign_related_fields
    return tuple(rhs_field for lhs_field, rhs_field in self.related_fields if rhs_field)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/related.py", line 621, in related_fields
    self._related_fields = self.resolve_related_fields()
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/related.py", line 606, in resolve_related_fields
    raise ValueError('Related model %r cannot be resolved' % self.remote_field.model)
ValueError: Related model 'booking.Calendar' cannot be resolved

Upon inspection, this was due to the initial migrations for the `booking` app not having been created yet.  The `booking` app is where the non-abstract `Calendar` lives. 

However, trying to create the initial migrations for the app results in pointing back to the `scheduler` migrations.

>   File "/usr/local/lib/python3.6/site-packages/django/db/migrations/operations/fields.py", line 196, in state_forwards
    state.models[app_label, self.model_name_lower].fields
KeyError: ('schedule', 'calendarrelation')

**Comment by thongly** [comment]
Perhaps splitting up the app might help alleviate the migration issues - [as in this PR for vera.](https://github.com/wq/vera/commit/12501471eb24e15425dc93e3c6930d331424e21e)

I'm not familiar enough with the swappable migrations, but I'll try to fix based on this -
 https://github.com/wq/django-swappable-models#migration-scripts

File: schedule/models/calendars.py [hunk 9]
```diff
@@ -218,7 +228,7 @@ class CalendarRelation(models.Model):
     may not scale well.  If you use this, keep that in mind.
     '''
 
-    calendar = models.ForeignKey(Calendar, on_delete=models.CASCADE, verbose_name=_("calendar"))
+    calendar = models.ForeignKey(SCHEDULER_CALENDAR_MODEL, on_delete=models.CASCADE, verbose_name=_("calendar"))
```
Comment (mikekeda) [inline schedule/models/calendars.py]: Line too long https://www.python.org/dev/peps/pep-0008/#maximum-line-length

File: schedule/models/calendars.py [hunk 10]
```diff
@@ -182,6 +154,44 @@ def get_absolute_url(self):
...
 
 
+class Calendar(AbstractCalendar):
+    """
+    This is for grouping events so that batch relations can be made to all
+    events.  An example would be a project calendar.
+
+    name: the name of the calendar
+    events: all the events contained within the calendar.
+    >>> calendar = Calendar(name = 'Test Calendar')
```
Comment (mikekeda) [inline schedule/models/calendars.py]: `calendar = Calendar(name='Test Calendar')`
Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value. https://www.python.org/dev/peps/pep-0008/#other-recommendations

File: schedule/models/calendars.py [hunk 11]
```diff
@@ -104,46 +104,18 @@ def get_calendars_for_object(self, obj, distinction=''):
 
 
 @python_2_unicode_compatible
-class Calendar(models.Model):
+class AbstractCalendar(models.Model):
     '''
```
Comment (mikekeda) [inline schedule/models/calendars.py]: I know it's not your line, but could you please replace ''' with """ in docstring?
https://www.python.org/dev/peps/pep-0008/#documentation-strings

File: schedule/models/events.py [hunk 12]
```diff
@@ -47,10 +47,10 @@ def get_for_object(self, content_object, distinction='', inherit=True):
 
 
 @python_2_unicode_compatible
-class Event(models.Model):
+class AbstractEvent(models.Model):
     '''
```
Comment (mikekeda) [inline schedule/models/events.py]: I know it's not your line, but could you please replace ''' with """ in docstring?
https://www.python.org/dev/peps/pep-0008/#documentation-strings

File: schedule/utils.py [hunk 13]
```diff
@@ -104,29 +105,26 @@ def get_kwarg_or_param(request, kwargs, key):
...
     else:
         event_id = get_kwarg_or_param(request, kwargs, 'event_id')
-        event = Event.objects.filter(pk=event_id).first() if event_id else None
+        event = get_event_model().objects.filter(pk=event_id).first() if event_id else None
     return event
 
 
 def get_calendar(event, request, **kwargs):
-    from schedule.models import Calendar
     calendar = None
```
Comment (mikekeda) [inline schedule/utils.py]: It's not your line, but could you please remove it? I think it's pointless.

**Comment by llazzaro** [comment]
Thank you everyone for comments and reviews on this pull request.
I will soon try to review this PR.
@jdufresne are you ok with merging this changes? I didn't have time to review them, but if it's ok and you too... I will merge it soon. If not I will comeback here to add more comments

**Comment by jdufresne** [comment]
Sorry, @llazzaro. I won't have time in the near future to review this. If it looks good to you, please proceed.

**Comment by lggwettmann** [comment]
Is this going to merge any time soon? (Thanks for the great work, everyone!)

**Comment by lggwettmann** [comment]
What can I do to move this forward? I need to customize the occurrence model and would prefer to base that on the common project code. :)

**Comment by pickyuptruck** [comment]
any updates on this?

**Comment by m-vdb** [comment]
I am no longer using `django-scheduler`, so I think it might be best if someone took over :)
